File: 2001/5/8-5b.py
#!/usr/bin/env python

import sys

mem = []
extra = 0
for line in sys.stdin:
    if line == "":
        break
    if line[-1] == "\n":
        line = line[:-1]
    cont = 0
    for i in xrange(len(line)):
        ch = line[i]
        if cont:
            cont -= 1
            continue
        if ord(ch) >= ord('0') and ord(ch) <= ord('9'):
            sys.stdout.write(ch)
            while len(mem) and ord(line[i + 1]) >= ord('0') and ord(line[i + 1]) <= ord('9'):
                sys.stdout.write(' ' + mem.pop(0) + ' ' + line[i + 1] + ")")
                i += 1
                cont += 1
            if len(mem) and i < len(line):
                sys.stdout.write(mem.pop(0))
                extra += 1
        else:
            sys.stdout.write("(")
            mem.append(ch)
    for i in xrange(extra):
        sys.stdout.write(")")
    print

File: 2001/5/8-5b.cpp
#include <iostream>
#include <string>
#include <stack>
using namespace std;

int main()
{
    string st;
    char tmp;
    stack<char> mem;
    int extra = 0;
    while (cin >> st)
    {
        for (int i = 0; i < st.length(); i++)
        {
            tmp = st.at(i);
            if (!isdigit(tmp))
            {
                cout << "(";
                mem.push(tmp);
            }
            else
            {
                cout << tmp;
                while ((!mem.empty()) && (isdigit(st[i + 1])))
                {
                    cout << " " << mem.top() << " " << st[i + 1] << ")";
                    mem.pop();
                    i++;
                }
                if ((!mem.empty()) && (i < st.length()))
                {
                    cout << " " << mem.top() << " ";
                    mem.pop();
                    extra++;
                }
            }
        }
        for (extra; extra > 0; extra--)
            cout << ")";
        cout << endl;
    }
}
